arangod/RocksDBEngine/RocksDBPrimaryIndex.cpp

```diff
@@ -425,42 +550,224 @@ IndexIterator* RocksDBPrimaryIndex::iteratorForCondition(
   TRI_ASSERT(node->type == aql::NODE_TYPE_OPERATOR_NARY_AND);
-  TRI_ASSERT(node->numMembers() == 1);
+  TRI_ASSERT(node->numMembers() >= 1);
+  TRI_ASSERT(node->numMembers() <= 2);
+
+  // LOG_DEVEL << "iteratorForCondition";
+  // LOG_DEVEL << arangodb::aql::AstNode::toString(node);
+
+  static std::string const& lowest(StaticStrings::Empty);
+  static std::string const hightest = "\xFF";
+
+  static const auto removeCollectionFromString =
+      [this, &trx](bool isId, std::string& value) -> bool {
+    if (isId) {
+      char const* key = nullptr;
+      size_t outLength = 0;
+      std::shared_ptr<LogicalCollection> collection;
+      Result res = trx->resolveId(value.data(), value.length(), collection, key, outLength);
+
+      if (!res.ok()) {
+        return false;
+      }
+
+      TRI_ASSERT(key);
+      TRI_ASSERT(collection);
+
+      if (!_isRunningInCluster && collection->id() != _collection.id()) {
+        // only continue lookup if the id value is syntactically correct and
+        // refers to "our" collection, using local collection id
+        return false;
+      }
+
+      if (_isRunningInCluster) {
+        if (collection->planId() != _collection.planId()) {
+          // only continue lookup if the id value is syntactically correct and
+          // refers to "our" collection, using cluster collection id
+          return false;
+        }
+      }
+
+      value = std::string(key, outLength);
+    }
+    return true;
+  };
+
+  if (node->numMembers() == 1) {
+    auto comp = node->getMember(0);
+    // assume a.b == value
+    auto attrNode = comp->getMember(0);
+    auto valNode = comp->getMember(1);
+
+    if (attrNode->type != aql::NODE_TYPE_ATTRIBUTE_ACCESS) {
+      // value == a.b  ->  flip the two sides
+      attrNode = comp->getMember(1);
+      valNode = comp->getMember(0);
+    }
 
-  auto comp = node->getMember(0);
-  // assume a.b == value
-  auto attrNode = comp->getMember(0);
-  auto valNode = comp->getMember(1);
+    TRI_ASSERT(attrNode->type == aql::NODE_TYPE_ATTRIBUTE_ACCESS);
+
+    if (comp->type == aql::NODE_TYPE_OPERATOR_BINARY_EQ) {
+      // a.b == value
+      return createEqIterator(trx, attrNode, valNode);
+    }
 
-  if (attrNode->type != aql::NODE_TYPE_ATTRIBUTE_ACCESS) {
-    // value == a.b  ->  flip the two sides
-    attrNode = comp->getMember(1);
-    valNode = comp->getMember(0);
+    if (comp->type == aql::NODE_TYPE_OPERATOR_BINARY_IN) {
+      // a.b IN values
+      if (valNode->isArray()) {
+        // a.b IN array
+        return createInIterator(trx, attrNode, valNode);
+      }
+    }
   }
-  TRI_ASSERT(attrNode->type == aql::NODE_TYPE_ATTRIBUTE_ACCESS);
 
-  if (comp->type == aql::NODE_TYPE_OPERATOR_BINARY_EQ) {
-    // a.b == value
-    return createEqIterator(trx, attrNode, valNode);
+  std::unique_ptr<std::string> lower = nullptr;
+  std::unique_ptr<std::string> upper = nullptr;
+
+  // add nodes that may create an range
+  std::vector<aql::AstNode const*> nodes;
```

Comment:
don't we always have at most 2 nodes here?
why not use an std::array of size 2 then? that would save any memory allocations.